docs/regression_test_api.rst

```diff
@@ -36,71 +36,97 @@ Pipeline Hooks
 .. autodecorator:: reframe.core.decorators.require_deps
 
 
-.. _directives:
 
-Directives
-----------
+Builtins
+--------
 
-Directives are functions that can be called directly in the body of a ReFrame regression test class. 
-These functions exert control over the test creation, and they allow adding and/or modifying certain attributes of the regression test.
-For example, a test can be parameterized using the :func:`parameter` directive as follows:
+ReFrame provides built-in functions that facilitate the creation of extensible tests (i.e. a test library).
+These *builtins* are intended to be used directly in the class body of the test, allowing the ReFrame internals to *pre-process* their input before the actual test creation takes place.
+This provides the ReFrame internals with further control over the user's input, making the process of writing regression tests less error-prone thanks to a better error checking.
+In essence, these builtins exert control over the test creation, and they allow adding and/or modifying certain attributes of the regression test.
 
-.. code:: python
+.. py:function:: reframe.core.pipeline.RegressionTest.parameter(values=None, inherit_params=False, filter_params=None)
+
+  Inserts or modifies a regression test parameter.
```

Comment:
In the documentation, I am missing a bit -as a user- the connection with the `@parameterized_test`. What are the differences and when should I use this syntax over the other. We recently came across an example with @ekouts that the `@parameterized_test` would be at least as useful:

```python
@rfm.parameterized_test([1, 2], [3, 4])
class X(rfm.RegressionTest):
    def __init__(self, a, b):
       pass
```

With the parameter syntax, I think that you have to combine the two parameters into one, since we are only looking for two points in the parameter space:

```python
class X(rfm.RegressionTest):
    ab = parameter([1, 2], [3, 4])
    def __init__(self):
        self.a, self.b = self.ab
```

Is a better way to write this with the `parameter()` built-in ?

Reply:
The naming should certainly change, but let's not do in this PR. The current naming scheme will work even for lists, but it will produce a very ugly output, something like this for this example: `X_1__2_` when `ab = [1, 2]` 😬

Reply:
I suggest opening an issue not forget about updating the naming scheme. I believe that all these changes will bring us to 4.0 :-)